torch/csrc/jit/interpreter.cpp

```diff
@@ -1001,14 +1003,24 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
             ++af.pc;
           } break;
           case GUARD: {
-            auto actual = TensorType::create(stack.back().toTensor());
-            const TypePtr& expected = af.types[inst.X];
-            push(stack, *expected == *actual);
+            // TODO: In future, `expected` should probably work as a mask
+            // i.e. if some dimension in `expected` is set to `*`, it should
+            // accept any dimension in `actual`
+            // Similarly, if datatype is set to a wildcard, `GUARD` should
+            // accept any datatype
+            auto t = stack.back().toTensor();
+            const TypePtr &expected = af.types[inst.X];
+            if (t.defined()) {
+              auto actual = TensorType::create(stack.back().toTensor());
+              push(stack, *expected == *actual);
+            } else {
+              push(stack, expected->expect<TensorType>()->autogradZero());
```

Comment:
I believe this can push None and not a bool and then the if statement will crash. 

tensor is undefined and autogradZero in guard is:
None (is i.e. either is ok): true
true (i.e. it needs to be autograd zero): true
false: false

Reply:
this whole logic becomes just `push(stack, *expected == *actual);` in autodiff PR. I'll just remove this snippet.